### PR TITLE
PERF-XDP-TCP-NTTTCP: New testcase to verify xdp throughput performance

### PR DESCRIPTION
**Pull request by abhimarathe**
1. Introduce new testcase PERF-XDP-TCP-NTTTCP: To measure throughput overhead introduced by XDP.
2. Enable custom kernel install from tar source file.
 Now you can use tar source file to install kernel like shown below:
`.\Run-LisaV2.ps1 -TestLocation 'eastus2' -RGIdentifier 'abmarathTest' -TestPlatform 'Azure' -ARMImageName 'Canonical UbuntuServer 18.04-LTS latest' -TestNames 'PERF-XDP-TCP-NTTTCP' -ResourceCleanup keep -XMLSecretFile "C:\Users\ABMARATH\Documents\Work\LSG-Secret\AzureSecretsTESTONLY.xml" -CustomKernel 'https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/snapshot/linux-5.7.tar.gz'`

I have ran this test over 3 connections just to quickly get the results..
```
[LISAv2 Test Results Summary]
Test Run On           : 08/12/2020 23:22:37
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Initial Kernel Version: 5.3.0-1034-azure
Final Kernel Version  : 5.7.0
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:1:5

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  PERF-XDP-TCP-NTTTCP                                                               PASS                48.53
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, setupType: TwoVM2NIC, TestLocation: eastus2
      Without XDP Throughput 1 connections : 30.35
      With XDP Throughput 1 connections : 30.39
      Without XDP Throughput 2 connections : 30.45
      With XDP Throughput 2 connections : 30.45
      Without XDP Throughput 4 connections : 30.45
      With XDP Throughput 4 connections : 30.45



Logs can be found at C:\LISAv2\SB51\TestResults\2020-08-12-16-21-57-3318
```

### CODE REVIEW

File: Testscripts/Linux/perf_ntttcp.sh [hunk 1]
```diff
@@ -215,7 +220,7 @@ Get_VFName()
 	# -K changed into --show-nic-packets, -I changed into --show-dev-interrupts in master branch
 	if [ $currentVersion ] && ( [ $currentVersion \> "1.4.0" ] || [ $currentVersion == "1.4.0" ] || [ $currentVersion == "master" ] ); then
 		vf_interface=$(ssh "$ip" lshw -c network -businfo | grep -i "Virtual Function" | awk '{print $2}')
-		if [ $vf_interface ]; then
+		if [[ $vf_interface ]]; then
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/perf_ntttcp.sh:223]: just curious if there are multiple VF interfaces.

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    $beforeDirPath = Join-Path -Path $args[0] -ChildPath 'report.csv'
+    $afterDirPath = Join-Path -Path $args[1] -ChildPath 'report.csv'
+
+    Write-LogInfo "Importing CSV from path $afterDirPath"
+    $csvWithXDP = Import-Csv -Path $afterDirPath
+    Write-LogInfo "Importing CSV from path $beforePath"
+    $csvWithoutXDP = Import-Csv -Path $beforeDirPath
+    $testFailedFlag = $false
+    if ( ($csvWithXDP).Count -ne ($csvWithoutXDP).Count) {
+        Throw "NTTTCP did not executed properly. Different num lines in report.csv"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/PERF-XDP-NTTTCP.ps1]: Please add counts data. Think if you have the test failure and read the log. The logging should be helpful.

**Review by juhlee-microsoft — COMMENTED** [review]
Please follow up my comment

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 (hunk 2 above)
Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-NTTTCP.ps1]: Fixed in: 52245c7dce4934e96ef35dcc4c6aa36315255e62

**Comment by abhimarathe** [comment]
> Please follow up my comment

Thank you Ju!
I have pushed new changes please take a look.

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        Write-LogInfo "  Internal IP : $($senderVMData.InternalIP)"
+
+        # PROVISION VMS FOR LISA WILL ENABLE ROOT USER AND WILL MAKE ENABLE PASSWORDLESS AUTHENTICATION ACROSS ALL VMS.
+        Provision-VMsForLisa -allVMData $allVMData -installPackagesOnRoleNames "none"
+
+        Write-LogInfo "Generating constants.sh ..."
+        $constantsFile = "$LogDir\constants.sh"
+        Set-Content -Value "#Generated by Azure Automation." -Path $constantsFile
+        Add-Content -Value "ip=$($receiverVMData.InternalIP)" -Path $constantsFile
+        Add-Content -Value "client=$($receiverVMData.InternalIP)" -Path $constantsFile
```
Comment (harz566) [inline Testscripts/Windows/PERF-XDP-NTTTCP.ps1:153]: IMO, receiver is server, sender is client, why it was changed as opposite here?

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        Add-Content -Value "server=$($senderVMData.InternalIP)" -Path $constantsFile
+        Add-Content -Value "testServerIP=$($senderVMData.SecondInternalIP)" -Path $constantsFile
```
Comment (harz566) [inline Testscripts/Windows/PERF-XDP-NTTTCP.ps1]: it's a little bit complicated design, especially to change the ntttcp_perf.sh, which should be a stable and common file to be used by wide test cases.   Why not enable/disable XDP from one of server/sender and then Run ntttcp again, and compare the result?

File: Testscripts/Linux/customKernelInstall.sh [hunk 5]
```diff
@@ -431,10 +432,14 @@ function InstallKernel() {
         fi
     fi
     if [[ ${CustomKernel} == "linuxnext" ]] || [[ ${CustomKernel} == "netnext" ]] || \
-        [[ ${CustomKernel} == "upstream-stable" ]]; then
+        [[ ${CustomKernel} == "upstream-stable" ]] || [[ ${CustomKernel} == *.tar.gz ]]; then
         LogMsg "Custom Kernel:$CustomKernel"
         Install_Build_Deps
-        sourceDir=$(Get_Upstream_Source "." "$kernelSource")
+        if [[ ${CustomKernel} == *.tar.gz ]]; then
```
Comment (harz566) [inline Testscripts/Linux/customKernelInstall.sh]: why here only focus on tar.gz,  what about *.tar  only? Can it be merged with line 215?

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 (hunk 3 above)
Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-NTTTCP.ps1:153]: It really doesn't matter. IMO Client is going to run XDP and XDP runs on RX queues so I always tends to select Client for receiver.

File: Testscripts/Linux/customKernelInstall.sh (hunk 5 above)
Comment (abhimarathe) [reply Testscripts/Linux/customKernelInstall.sh]: fixed

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 (hunk 4 above)
Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-NTTTCP.ps1]: As discussed over teams call. I have made changes.

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/18/2020 00:00:29
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:36

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  PERF-XDP-TCP-NTTTCP                                                               PASS                35.31
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.7.0
      Without XDP Throughput 1 connections : 20.12
      With XDP Throughput 1 connections : 21.62
      Without XDP Throughput 2 connections : 30.43
      With XDP Throughput 2 connections : 30.37


Logs can be found at C:\LISAv2\HA58\TestResults\2020-08-17-16-59-43-7665
```

**Review by harz566 — DISMISSED** [review]

File: XML/TestCases/PerformanceTests.xml [hunk 6]
```diff
@@ -890,6 +890,29 @@
       <Networking>SRIOV</Networking>
     </SetupConfig>
   </test>
+  <test>
+    <testName>PERF-XDP-TCP-NTTTCP</testName>
+    <testScript>PERF-XDP-NTTTCP.ps1</testScript>
+    <files>.\Testscripts\Linux\utils.sh,.\Testscripts\Linux\XDPDumpSetup.sh,.\Testscripts\Linux\perf_ntttcp.sh</files>
+    <TestParameters>
+        <param>testDuration=NTTTCP_RUNTIME_IN_SECONDS</param>
```
Comment (juhlee-microsoft) [inline XML/TestCases/PerformanceTests.xml:920]: Are those parameters changed frequently?

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+<#
+.Description
+    This script deploys the VM and Verifies there are no regression in Network Performance due
+    to XDP by comparing ntttcp results.
+#>
+
+param([object] $AllVmData,
+    [object] $CurrentTestData)
+
+$iFaceName = "eth1"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/PERF-XDP-NTTTCP.ps1:13]: You can pick up the interface name from the system. Please check out the old PR.

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        elseif ($currentState -imatch "TestSkipped") {
+            Write-LogErr "Test Skipped. Last known status: $currentStatus"
+            $testResult = "SKIPPED"
+        }
+        elseif ($currentState -imatch "TestFailed") {
+            Write-LogErr "Test failed. Last known status: $currentStatus."
+            $testResult = "FAIL"
+        }
+        else {
+            Write-LogErr "Test execution is not successful, check test logs in VM."
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/PERF-XDP-NTTTCP.ps1]: It is nice to post the $currentState in the log even blank

File: Testscripts/Linux/customKernelInstall.sh [hunk 9]
```diff
@@ -17,7 +17,7 @@
 
 supported_kernels=(ppa proposed proposed-azure proposed-edge latest
                     linuxnext netnext upstream-stable)
-
+tarDestination="./linux-source"
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/customKernelInstall.sh:20]: Where does it come from?

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: XML/TestCases/PerformanceTests.xml (hunk 6 above)
Comment (abhimarathe) [reply XML/TestCases/PerformanceTests.xml:920]: These parameters are kept at user configuration level as other NTTTCP tests are keeping it.

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 (hunk 7 above)
Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-NTTTCP.ps1:13]: I'm going to follow-up on this with other PR as all XDP testcases needs to updated.

File: Testscripts/Linux/customKernelInstall.sh (hunk 9 above)
Comment (abhimarathe) [reply Testscripts/Linux/customKernelInstall.sh:20]: I included it to keep a common destination after tar file is extracted. Why?

**Review by harz566 — DISMISSED** [review]

**Review by juhlee-microsoft — DISMISSED** [review]

**Comment by johnsongeorge-w** [comment]
As discussed please check for storing the performance data.

**Review by johnsongeorge-w — COMMENTED** [review]
As discussed please check for storing the performance data.

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/24/2020 21:40:47
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:32

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  PERF-XDP-TCP-NTTTCP                                                               PASS                30.39 
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.7.0
      Without XDP Throughput 1 connections : 22.52
      With XDP Throughput 1 connections : 23.05
      Without XDP Throughput 2 connections : 25.98
      With XDP Throughput 2 connections : 25.31
      


Logs can be found at C:\LISAv2\LM42\TestResults\2020-08-24-14-40-06-1107
```

I checked resultMap by printing it in logs and it is getting updated with correct values.

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+            Write-LogInfo "Current Test Status: $currentStatus"
+            Wait-Time -seconds 20
+            $timer += 1
+            if ($timer -gt 15) {
+                Throw "XDPSetup did not stop after 5 mins. Please check xdpConsoleLogs."
+            }
+        }
+
+        $currentState = Run-LinuxCmd -ip $receiverVMData.PublicIP -port $receiverVMData.SSHPort `
+            -username $user -password $password -command "cat ~/state.txt" -runAsSudo
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-XDP-NTTTCP.ps1]: state.txt exist under root?

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+        if ($currentState -imatch "Completed") {
+            # Start PERF test without XDP
+            $ResultDir = "$LogDir\WithoutXDP"
+            New-Item -Path $ResultDir -ItemType Directory -Force | Out-Null
+            Run_NTTTCP_PERF $ResultDir
+
+            # Start XDPDump on client
+            $xdp_command = "cd ~/bpf-samples/xdpdump && ./xdpdump -i $iFaceName > ~/xdpdumpoutPERF.txt"
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-XDP-NTTTCP.ps1:204]: please add turn off lro part, thanks!

Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-NTTTCP.ps1:204]: It will be handled by XDPDumpsetup.sh. We don't need to turn it off again as we are not rebooting or anything.
What do you think?

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 09/02/2020 20:52:51
ARM Image Under Test  : Canonical : 0001-com-ubuntu-server-focal : 20_04-lts : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:9

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              PERF-XDP-TCP-LATENCY                                                              PASS                 8.03 
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.4.0-1022-azure
      Without XDP Average Latency : 91.982
      With XDP Average Latency : 99.033
      


Logs can be found at C:\LISAv2\VD57\TestResults\2020-09-02-13-52-06-2379
[LISAv2 Test Results Summary]
Test Run On           : 09/02/2020 06:19:04
ARM Image Under Test  : Canonical : 0001-com-ubuntu-server-focal : 20_04-lts : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:53

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  PERF-XDP-TCP-NTTTCP                                                               PASS                48.91 
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_D64_v3, TestLocation: eastus2, Kernel Version: 5.4.0-1022-azure
      Without XDP Throughput 1 connections : 18.19
      With XDP Throughput 1 connections : 18.20
      Without XDP Throughput 2 connections : 30.41
      With XDP Throughput 2 connections : 30.42
      Without XDP Throughput 4 connections : 30.44
      With XDP Throughput 4 connections : 30.45
      


Logs can be found at C:\LISAv2\ZJ74\TestResults\2020-09-01-23-18-15-6501
```

File: Testscripts/Windows/PERF-XDP-NTTTCP.ps1 (hunk 10 above)
Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-NTTTCP.ps1]: fixed

File: Testscripts/Linux/perf_ntttcp.sh [hunk 12]
```diff
@@ -283,7 +282,7 @@ Run_Ntttcp()
 			tx_log_prefix="sender-${testType}-${bufferLength}k-p${num_threads_P}X${num_threads_n}.log"
 			rx_log_prefix="receiver-${testType}-${bufferLength}k-p${num_threads_P}X${num_threads_n}.log"
 			run_msg="Running ${testType} ${bufferLength}k Test: $current_test_threads connections : $num_threads_P X $num_threads_n X $client_count clients"
-			server_ntttcp_cmd="ulimit -n 204800 && ${ntttcp_cmd} -u -b ${bufferLength}k -P ${num_threads_P} -t ${testDuration} -e -W 1 -C 1"
+			server_ntttcp_cmd="ulimit -n 204800 && ${ntttcp_cmd} -r${server} -u -b ${bufferLength}k -P ${num_threads_P} -t ${testDuration} -e -W 1 -C 1"
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/perf_ntttcp.sh:285]: Why this change ?

File: Testscripts/Linux/perf_ntttcp.sh [hunk 13]
```diff
@@ -293,7 +292,7 @@ Run_Ntttcp()
 			tx_log_prefix="sender-${testType}-p${num_threads_P}X${num_threads_n}.log"
 			rx_log_prefix="receiver-${testType}-p${num_threads_P}X${num_threads_n}.log"
 			run_msg="Running ${testType} Test: $current_test_threads connections : $num_threads_P X $num_threads_n X $client_count clients"
-			server_ntttcp_cmd="ulimit -n 204800 && ${ntttcp_cmd} -P ${num_threads_P} -t ${testDuration} -e -W 1 -C 1"
+			server_ntttcp_cmd="ulimit -n 204800 && ${ntttcp_cmd} -r${server} -P ${num_threads_P} -t ${testDuration} -e -W 1 -C 1"
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/perf_ntttcp.sh:295]: Why this change ?

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/perf_ntttcp.sh (hunk 12 above)
Comment (abhimarathe) [reply Testscripts/Linux/perf_ntttcp.sh:285]: By default this script runs ntttcp on primary interface's IP. As I'm setting XDP on extra (Secondary) Interface, so that management traffic would be separate, added -r option.

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Linux/perf_ntttcp.sh (hunk 13 above)
Comment (abhimarathe) [reply Testscripts/Linux/perf_ntttcp.sh:295]: refer comments above.

**Review by johnsongeorge-w — DISMISSED** [review]
Check the ntttcp test without XDP for impact analysis.

**Review by juhlee-microsoft — APPROVED** [review]

**Review by johnsongeorge-w — APPROVED** [review]

**Review by harz566 — APPROVED** [review]
